### ISSUE TITLE
Export SSL certificates to /ssl

### ISSUE THREAD

**Issue by felipecrs** [brenner-tobias/addon-cloudflared#47]
It would be nice if the Cloudflared add-on could export the SSL certificates to /ssl like the DuckDNS and Let's Encrypt add-on does.

For me, this would be useful for other server add-ons like Asterisk, which requires a SSL certificate regardless.

I don't know if it's possible, but if it is, it would be nice.

If it is possible, I believe that perhaps we could recommend people setup their http integration to enable HTTPS, so that traffic inside of the house also gets encrypted.

**Comment by elcajon** [comment]
I am not sure if I get your question right, but the Cloudflared tunnel is working without any certificates for the communication between your server and the Cloudflare edge servers. So there is no way (and need) on creating/storing SSL certificates.

You can of course add SSL (HTTPS) support for all of your origin services to work encrypted on your LAN. This has no effect on the communication between cloudflared and Cloudflare edge servers.

**Comment by felipecrs** [comment]
> You can of course add SSL (HTTPS) support for all of your origin services to work encrypted on your LAN. This has no effect on the communication between cloudflared and Cloudflare edge servers.

And in order to add SSL support for all my origin services, I need the certificates. I know that Cloudflare can provide the origin certificates for free as well:

https://developers.cloudflare.com/ssl/origin-configuration/origin-ca

So that, the proposal is to add a feature in the add-on, that, if enabled, would generate/retrieve the origin SSL certificates from Cloudflare and place them at `/ssl` so that I can configure my Home Assistant to use it.

For the domain names, I suspect the add-on should instruct the Cloudflare API to generate a certificate which is valid for both the main domain (for Home Assistant) and any additional hosts specified in the configuration.

An example of an add-on which would require it in order to fully function is the TasmoAdmin, when combined with the sidepanel.

**Comment by felipecrs** [comment]
References:

https://developers.cloudflare.com/ssl/origin-configuration/origin-ca#api-calls
https://blog.cloudflare.com/cloudflare-ca-encryption-origin/

**Comment by elcajon** [comment]
Alright, now I get your point. I thought the request was related directly to cloudflared. 
I think @brenner-tobias has to make the basic decision whether the general Cloudflare API should be supported in the add-on in addition to the cloudflared tunnel. That's beyond my competencies. 😉

**Comment by brenner-tobias** [comment]
Thanks both of you for raising this. Generally speaking, I do support this 100%. A couple of remarks from my side to make sure we are on the same page regarding the functionality and the benefits here:

We could create Origin Certificates for the different hosts. This would mean, that we could create one for HA itself and for the other services in `additional_hosts`. This would mean, that these certificates could be stored in the SSL folder and then moved from there to the individual services manually. Now Cloudflare specifically states that 

> Origin Certificates are only valid for encryption between Cloudflare and your origin server.

Of course, if you are in a private network where you call the IP, certificates are never really valid, so that might not be a problem. But it is also not ideal for the local networks. 

So there are two benefits:

1. Activating SSL for HA to also use https in the local network without any manual steps.
2. Streamlining the process to add certificates for additional_hosts in an easy way.

Regarding benefit 1, I think there is no doubt this would be helpful. Regarding #2, I am not sure this is really easier then going on the Cloudflare page and creating the certificates yourself. You would need to ssh into HA and copy the certificate files to your server, which is kind of harder than going to the website. Also this would limit you to not work with an existing private key and CSR.

So to summarise: Let's give it a shot for implementing the download of the certificate for HA as a first step. We can then decide if we go ahead and also implement something for the additional_hosts.

Does that make sense?

**Comment by felipecrs** [comment]
>  Regarding 2, I am not sure this is really easier then going on the Cloudflare page and creating the certificates yourself. You would need to ssh into HA and copy the certificate files to your server, which is kind of harder than going to the website. Also this would limit you to not work with an existing private key and CSR.

That's exactly why I believe it would be better if the add-on could generate a single certificate valid for all the hostnames.

It's even the default configuration when generating a certificate from Cloudflare:

> [The zone root and first level wildcard hostname are included by default.](https://developers.cloudflare.com/ssl/origin-configuration/origin-ca#:~:text=The%20zone%20root%20and%20first%20level%20wildcard%20hostname%20are%20included%20by%20default.)

**Comment by elcajon** [comment]
> Of course, if you are in a private network where you call the IP, certificates are never really valid, so that might not be a problem. But it is also not ideal for the local networks.

This is also my concern. 

I used the Cloudflare wildcard origin certificates back then when using NginxProxyManager with Cloudflare DNS and full (strict) SSL settings. For encrypting services inside my LAN any other selfsigned certificate is just as good as the Cloudflare origin certificate. I still have problems to see the benefit of collecting these zone certificates for hosts already tunneled through cloudflared?  But maybe I'm a little slow on the uptake today, so I don't want to slow you down.

**Comment by felipecrs** [comment]
One benefit about using the Cloudflare provided certificate is that it allows to enable strict SSL setting (not only for HA, but for any other additional server).

**Comment by NPC-112** [comment]
I ended up copying the private key and certificate from the cert.pem file to \ssl\privkey.pem and fullchan.pem, then adding them to the http: section of the home assistant config.
I have full (strict) enabled on cloudflare working no problem.

Only downside is I have to have the http: section filled and enabled across all of hass, making my local network "not secure" and not being able to use local network in the ios app.

Any reason why this should be avoided? Can or should I add these to my nginxproxymanager addon for my sub-hosts?

**Comment by brenner-tobias** [comment]
Strict SSL settings are already possible, though we are currently using the `noTLSVerify=true` option. This one can be removed. So I suggest we start working on it and see how far we want to take it.

**Comment by brenner-tobias** [comment]
> Agreed. I will start working on it.

Awesome, you can create a branch directly in this repo if you want, so we can collaborate on it
